man/u_man/man1/co.1
'\"macro stdmacro
.TH CO 1
.SH NAME
co \- check out RCS revisions
.SH SYNOPSIS
.B co
[\  options\ ] file ...
.SH DESCRIPTION
.I co
retrieves revisions from RCS files.
Each file name ending in `,v' is taken to be an RCS file.
All other files
are assumed to be working files.
\f2co\f1 retrieves a revision from each RCS file and stores it into
the corresponding working file.
.PP
Pairs of RCS files and working files may be specified in 3 ways (see also the
example section).
.PP
1) Both the RCS file and the working file are given. The RCS file name is of
the form \f2path1/workfile\f1,v
and the working file name is of the form
\f2path2/workfile\f1, where
\f2path1/\f1 and
\f2path2/\f1 are (possibly different or empty) paths and
\f2workfile\f1 is a file name.
.PP
2) Only the RCS file is given. Then the working file is created in the current
directory and its name is derived from the name of the RCS file
by removing \f2path1/\f1 and the suffix `,v'.
.PP
3) Only the working file is given.
Then the name of the RCS file is derived from the name of the working file
by removing \f2path2/\f1
and appending the suffix `,v'.
.PP
If the RCS file is omitted or specified without a path, then \f2co\f1
looks for the RCS file first in the directory \&./RCS and then in the current
directory.
.PP
Revisions of an RCS file may be checked out locked or unlocked. Locking a
revision prevents overlapping updates. A revision checked out for reading or
processing (e.g., compiling) need not be locked. A revision checked out
for editing and later checkin must normally be locked. Locking a revision
currently locked by another user fails. (A lock may be broken with
the \f2rcs\f1 (1) command.)
\f2co\f1 with locking requires the caller to be on the access list of
the RCS file, unless he is the owner of the
file or the superuser, or the access list is empty.
\f2co\f1 without locking is not subject to accesslist restrictions.
.PP
A revision is selected by number,
checkin date/time,
author, or state. If none of these options
are specified, the latest revision
on the trunk is retrieved.
When the options
are applied in combination, the latest revision
that satisfies all of them is retrieved.
The options for date/time, author, and 
state retrieve a revision on the \f2selected
branch\f1. The selected branch is either 
derived from the revision number (if given),
or is the highest branch on the trunk.
A revision number may be attached
to one of the options
\f3\-l\f1, \f3\-p\f1, \f3\-q\f1, \f3\-M\f1, or \f3\-r\f1.
.PP
A \f2co\f1 command applied to an RCS
file with no revisions creates a zero-length file.
\f2co\f1 always performs keyword substitution (see below).
.PP
.TP 11
.B \-l\f1[\f2rev\f1]
locks the checked out revision for the caller.
If omitted, the checked out revision is not locked.
See option \f3\-r\f1 for handling of the revision number \f2rev\f1.
.TP 11
.B \-p\f1[\f2rev\f1]
prints the retrieved revision on the std. output rather than storing it
in the working file.
This option is useful when \f2co\f1
is part of a pipe.
.TP 11
.B \-q\f1[\f2rev\f1]
quiet mode; diagnostics are not printed.
.TP 11
. B \-M\f1[\f2rev\f1]
the checked out file's last-modified date is set to the revision date
instead of the current date.
.TP 11
.BI \-d "date"
retrieves the latest revision on the selected branch whose checkin date/time is less than or equal to \f2date\f1.
The date and time may be given in free format and are converted to local time.
Examples of formats for \f2date\f1:
.sp 1
.nf
	\f222-April-1982, 17:20-CDT,
	2:25 AM, Dec. 29, 1983,
	Tue-PDT, 1981, 4pm Jul 21\f1         \f1(free format),
	\f2Fri, April 16 15:52:25 EST 1982 \f1(output of ctime).
.fi
.sp1
Most fields in the date and time may be defaulted.
\f2co\f1 determines the defaults in the order year, month, day,
hour, minute, and second (most to least significant). At least one of these
fields must be provided. For omitted fields that are of higher significance
than the highest provided field,
the current values are assumed. For all other omitted fields,
the lowest possible values are assumed.
For example, the date "20, 10:30" defaults to
10:30:00 of the 20th of the current month and current year.
The date/time must be quoted if it contains spaces.
.TP 11
.B \-r\f1[\f2rev\f1]
retrieves the latest revision whose number is less than or equal to \f2rev\f1.
If \f2rev\f1 indicates a branch rather than a revision,
the latest revision on that branch is retrieved.
\f2Rev\f1 is composed of one or more numeric or symbolic fields
separated by `.'. The numeric equivalent of a symbolic field
is specified with the \f3\-n\f1 option of the commands \f2ci\f1 and \f2rcs\f1.
.TP 11
.BI \-s "state"
retrieves the latest revision on the selected branch whose state is set to \f2state\f1.
.TP 11
.B \-w\f1[\f2login\f1]
retrieves the latest revision on the selected branch which was checked in
by the user with login name \f2login\f1. If the argument \f2login\f1 is
omitted, the caller's login is assumed.
.TP 11
.B \-j\f2joinlist\f1
generates a new revision which is the join of the revisions on \f2joinlist\f1.
\f2Joinlist\f1 is a comma-separated list of pairs of the form
\f2rev2:rev3\f1, where \f2rev2\f1 and \f2rev3\f1 are (symbolic or numeric)
revision numbers.
For the initial such pair, \f2rev1\f1 denotes the revision selected
by the options \f3\-l\f1, ..., \f3\-w\f1. For all other pairs, \f2rev1\f1
denotes the revision generated by the previous pair. (Thus, the output
of one join becomes the input to the next.)
.sp
For each pair, \f2co\f1 joins revisions \f2rev1\f1 and \f2rev3\f1
with respect to \f2rev2\f1.
This means that all changes that transform
\f2rev2\f1 into \f2rev1\f1 are applied to a copy of \f2rev3\f1.
This is particularly useful if \f2rev1\f1
and \f2rev3\f1 are the ends of two branches that have \f2rev2\f1 as a common
ancestor. If \f2rev1\f1 < \f2rev2\f1 < \f2rev3\f1 on the same branch,
joining generates a new revision which is like \f2rev3\f1, but with all
changes that lead from \f2rev1\f1 to \f2rev2\f1 undone.
If changes from \f2rev2\f1 to \f2rev1\f1 overlap with changes from
\f2rev2\f1 to \f2rev3\f1, \f2co\f1 prints a warning and includes the
overlapping sections, delimited by the lines \f2<<<<<<<\ rev1,
=======\f1, and \f2>>>>>>>\ rev3\f1.
.PP
For the initial pair, \f2rev2\f1 may be omitted. The default is the common
ancestor.
If any of the arguments indicate branches, the latest revisions
on those branches are assumed. If the option \f3\-l\f1 is present,
the initial \f2rev1\f1 is locked.
.SH "KEYWORD SUBSTITUTION"
Strings of the form \f2$keyword$\f1 and \f2$keyword:...$\f1 embedded in
the text are replaced
with strings of the form \f2$keyword:\ value\ $\f1,
where \f2keyword\f1 and \f2value\f1 are pairs listed below.
Keywords may be embedded in literal strings
or comments to identify a revision.
.PP
Initially, the user enters strings of the form \f2$keyword$\f1.
On checkout, \f2co\f1 replaces these strings with strings of the form
\f2$keyword:\ value\ $\f1. If a revision containing strings of the latter form
is checked back in, the value fields will be replaced during the next
checkout.
Thus, the keyword values are automatically updated on checkout.
.PP
Keywords and their corresponding values:
.TP 13
$\&Author$
The login name of the user who checked in the revision.
.TP
$\&Class$
Prog, Def, Doc, or Test, depending on the class assigned to the file
with the \f3-c\f1 option of the \f2rcs\f1 command.
.TP
$\&Date$
The date and time the revision was checked in.
.TP
$\&Header$
A standard header containing the RCS file name, the
revision number, the date, the author, and the state.
.TP
$\&Locker$
The login name of the user who locked the revision (empty if not locked).
.TP
$\&Log$
The log message supplied during checkin, preceded by a header
containing the RCS file name, the revision number, the author, and the date.
Existing log messages are NOT replaced.
Instead, the new log message is inserted after \f2$\&Log:...$\f1.
This is useful for
accumulating a complete change log in a source file.
.TP
$\&Revision$
The revision number assigned to the revision.
.TP
$\&Source$
The full pathname of the RCS file.
.TP
$\&State$
The state assigned to the revision with \f2rcs\f3 \-s\f1 or \f2ci\f3 \-s\f1.
.SH DIAGNOSTICS
The RCS file name, the working file name,
and the revision number retrieved are
written to the diagnostic output.
The exit status always refers to the last file checked out,
and is 0 if the operation was successful, 1 otherwise.
.SH EXAMPLES
Suppose the current directory contains a subdirectory `RCS' with an RCS file
`io.c,v'. Then all of the following commands retrieve the latest
revision from `RCS/io.c,v' and store it into `io.c'.
.nf
.sp
        co  io.c;    co RCS/io.c,v;    co  io.c,v;
        co  io.c  RCS/io.c,v;    co  io.c  io.c,v;
        co  RCS/io.c,v  io.c;    co  io.c,v  io.c;
.fi
.SH "FILE MODES"
The working file inherits the read and execute permissions from the RCS
file. In addition, the owner write permission is turned on, unless the file
is checked out unlocked and locking is set to \f2strict\f1 (see
\f2rcs\f1 (1)).
.PP
If a file with the name of the working file exists already and has write
permission, \f2co\f1 aborts the checkout if \f3\-q\f1 is given, or asks
whether to abort if \f3\-q\f1 is not given. If the existing working file is
not writable, it is deleted before the checkout.
.SH FILES
The caller of the command must have write permission in the working
directory, read permission for the RCS file, and either read permission
(for reading) or read/write permission (for locking) in the directory which
contains the RCS file.
.PP
A number of temporary files are created.
A semaphore file is created in the directory of the RCS file
to prevent simultaneous update.
.SH IDENTIFICATION
Author: Walter F. Tichy,
.br
Purdue University, West Lafayette, IN, 47907.
.br
Copyright \(co 1982 by Walter F. Tichy.
.SH SEE ALSO
.nf
ci(1), ident(1), rcs(1), rcsdiff(1), rcsintro(1), rcsmerge(1), rlog(1), rcsfile(4)
.fi
.sp 0
Walter F. Tichy, ``Design, Implementation, and Evaluation of a Revision Control
System,'' in \f2Proceedings of the 6th International Conference on Software
Engineering\f1, IEEE, Tokyo, Sept. 1982.
.SH LIMITATIONS
The option \f3\-d\f1 gets confused in some circumstances,
and accepts no date before 1970.
There is no way to suppress the expansion of keywords, except
by writing them differently. In nroff and troff, this is done by embedding the
null-character `\\&' into the keyword.
.SH BUGS
The option \f3\-j\f1 does not work for
files that contain lines with a single `.'.


usr.bin/rcs/src/rcs/co.c
/*
 *                     RCS checkout operation
 */
 static char rcsid[]=
 "$Header: /d2/3.7/src/usr.bin/rcs/src/rcs/RCS/co.c,v 1.1 89/03/27 18:21:36 root Exp $ Purdue CS";
/*****************************************************************************
 *                       check out revisions from RCS files
 *****************************************************************************
 *
 * Copyright (C) 1982 by Walter F. Tichy
 *                       Purdue University
 *                       Computer Science Department
 *                       West Lafayette, IN 47907
 *
 * All rights reserved. No part of this software may be sold or distributed
 * in any form or by any means without the prior written permission of the
 * author.
 * Report problems and direct all inquiries to Tichy@purdue (ARPA net).
 */


/* Log:	co.c,v 
 * Revision 3.7  83/02/15  15:27:07  wft
 * Added call to fastcopy() to copy remainder of RCS file.
 * 
 * Revision 3.6  83/01/15  14:37:50  wft
 * Added ignoring of interrupts while RCS file is renamed; this avoids
 * deletion of RCS files during the unlink/link window.
 *
 * Revision 3.5  82/12/08  21:40:11  wft
 * changed processing of -d to use DATEFORM; removed actual from
 * call to preparejoin; re-fixed printing of done at the end.
 *
 * Revision 3.4  82/12/04  18:40:00  wft
 * Replaced getdelta() with gettree(), SNOOPDIR with SNOOPFILE.
 * Fixed printing of "done".
 *
 * Revision 3.3  82/11/28  22:23:11  wft
 * Replaced getlogin() with getpwuid(), flcose() with ffclose(),
 * %02d with %.2d, mode generation for working file with WORKMODE.
 * Fixed nil printing. Fixed -j combined with -l and -p, and exit
 * for non-existing revisions in preparejoin().
 *
 * Revision 3.2  82/10/18  20:47:21  wft
 * Mode of working file is now maintained even for co -l, but write permission
 * is removed.
 * The working file inherits its mode from the RCS file, plus write permission
 * for the owner. The write permission is not given if locking is strict and
 * co does not lock.
 * An existing working file without write permission is deleted automatically.
 * Otherwise, co asks (empty answer: abort co).
 * Call to getfullRCSname() added, check for write error added, call
 * for getlogin() fixed.
 *
 * Revision 3.1  82/10/13  16:01:30  wft
 * fixed type of variables receiving from getc() (char -> int).
 * removed unused variables.
 */




#include <pwd.h>
#include "rcsbase.h"
#include "time.h"
#include <sys/types.h>
#include <sys/stat.h>

static char rcsbaseid[] = RCSBASE;

extern FILE * fopen();
extern int    rename();
extern char * get_caller_name();
extern char * malloc();
extern struct hshentry * genrevs(); /*generate delta numbers                */
extern int  nextc;                  /*next input character                  */
extern int  nerror;                 /*counter for errors                    */
extern char * Kdesc;                /*keyword for description               */
extern char * maketempfile();       /*temporary file name                   */
extern char * buildrevision();      /*constructs desired revision           */
extern int    buildjoin();          /*join several revisions                */
extern char * mktempfile();         /*temporary file name generator         */
extern struct lock * addlock();     /*add a new lock                        */
extern long   maketime();           /*convert parsed time to unix time.     */
extern struct tm * localtime();     /*convert unixtime into a tm-structure  */
extern int StrictLocks;
extern FILE * finptr;               /* RCS input file                       */
extern FILE * frewrite;             /* new RCS file                         */

char * RCSfilename, * workfilename;
char * newRCSfilename, * neworkfilename;
int    rewriteflag; /* indicates whether input should be echoed to frewrite */
int usemodtime;			      /* use file rev time instead of cur*/

char * date, * rev, * state, * author, * join;
char finaldate[datelength];

int lockflag, tostdout;
char * caller;                        /* caller's login;                    */
extern quietflag;

char numericrev[revlength];           /* holds expanded revision number     */
struct hshentry * gendeltas[hshsize]; /* stores deltas to be generated      */
struct hshentry * targetdelta;        /* final delta to be generated        */

char * joinlist[joinlength];          /* pointers to revisions to be joined */
int lastjoin;                         /* index of last element in joinlist  */

main (argc, argv)
int argc;
char * argv[];
{
        register c;
        char * cmdusage;
        struct stat RCSstat;
        struct tm parseddate, *ftm;
        char * rawdate;
        long unixtime;
	int badfile;	/* true if error processing current file */

	nerror = 0;
	catchints();
        cmdid = "co";
        cmdusage = "command format:\nco -l[rev] -p[rev] -q[rev] -r[rev] -M[rev] -ddate -sstate -w[login] -jjoinlist file ...";
        date = rev = state = author = join = nil;
        lockflag = tostdout = quietflag = false;
	caller = get_caller_name();

        while (--argc,++argv, argc>=1 && ((*argv)[0] == '-')) {
                switch ((*argv)[1]) {

                case 'l':
                        lockflag=true;
                case 'r':
                revno:  if ((*argv)[2]!='\0') {
                                if (rev!=nil) warn("Redefinition of revision number");
                                rev = (*argv)+2;
                        }
                        break;

                case 'p':
                        tostdout=true;
                        goto revno;

                case 'q':
                        quietflag=true;
                        goto revno;

		case 'M':
			usemodtime=true;
			goto revno;

                case 'd':
                        if ((*argv)[2]!='\0') {
                                if (date!=nil) warn("Redefinition of -d option");
                                rawdate=(*argv)+2;
                        }
                        /* process date/time */
                        if (partime(rawdate,&parseddate)==0)
                                faterror("Can't parse date/time: %s",rawdate);
                        if ((unixtime=maketime(&parseddate))== 0L)
                                faterror("Inconsistent date/time: %s",rawdate);
                        ftm=localtime(&unixtime);
                        sprintf(finaldate,DATEFORM,
                        ftm->tm_year,ftm->tm_mon+1,ftm->tm_mday,ftm->tm_hour,ftm->tm_min,ftm->tm_sec);
                        date=finaldate;
                        break;

                case 'j':
                        if ((*argv)[2]!='\0'){
                                if (join!=nil)warn("Redefinition of -j option");
                                join = (*argv)+2;
                        }
                        break;

                case 's':
                        if ((*argv)[2]!='\0'){
                                if (state!=nil)warn("Redefinition of -s option");
                                state = (*argv)+2;
                        }
                        break;

                case 'w':
                        if (author!=nil)warn("Redefinition of -w option");
                        if ((*argv)[2]!='\0')
                                author = (*argv)+2;
                        else    author = caller;
                        break;

                default:
                        faterror("unknown option: %s\n%s", *argv,cmdusage);

                };
        } /* end of option processing */

        if (argc<1) faterror("No input file\n%s",cmdusage);

        /* now handle all filenames */
        do {
	badfile = 0;
        rewriteflag=false;
        finptr=frewrite=NULL;
        neworkfilename=nil;

        if (!pairfilenames(argc,argv,true,tostdout)) continue;

        /* now RCSfilename contains the name of the RCS file, and finptr
         * the file descriptor. If tostdout is false, workfilename contains
         * the name of the working file, otherwise undefined (not nil!).
         */
        diagnose("%s  -->  %s", RCSfilename,tostdout?"stdout":workfilename);

        fstat(fileno(finptr),&RCSstat); /* get file status, esp. the mode  */

        if (!tostdout && !trydiraccess(workfilename)) continue; /* give up */
        if (lockflag && !checkaccesslist(caller)) continue;     /* give up */
        if (!trysema(RCSfilename,lockflag)) continue;           /* give up */


        gettree();  /* reads in the delta tree */

        if (Head==nil) {
                /* no revisions; create empty file */
                diagnose("no revisions present; generating empty revision 0.0");
                if (!tostdout)
                        if (!creatempty(workfilename)) continue;
                else    putchar('\0'); /* end of file */
                /* Can't reserve a delta, so don't call addlock */
        } else {
                /* expand symbolic revision number */
                if (!expandsym(rev,numericrev))
                        continue;
                /* get numbers of deltas to be generated */
                if (!(targetdelta=genrevs(numericrev,date,author,state,gendeltas)))
                        continue;
                /* check reservations */
                if (lockflag && !addlock(targetdelta,caller,workfilename))
                        continue;

                if (join && !preparejoin()) continue;

                diagnose("revision %s %s",targetdelta->num,
                         lockflag?"(locked)":"");

                /* remove old working file if necessary */
                if (!tostdout)
                        if (!rmoldfile(workfilename)) continue;

                /* prepare for rewriting the RCS file */
                if (lockflag) {
                        newRCSfilename=mktempfile(RCSfilename,NEWRCSFILE);
                        if ((frewrite=fopen(newRCSfilename, "w"))==NULL) {
                                error("Can't open file %s",newRCSfilename);
				badfile++;
                                continue;
                        }
                        putadmin(frewrite);
                        puttree(Head,frewrite);
                        fprintf(frewrite, "\n\n%s%c",Kdesc,nextc);
                        rewriteflag=true;
                }

                /* skip description */
                getdesc(false); /* don't echo*/

                if (!(neworkfilename=buildrevision(gendeltas,targetdelta,
                      tostdout?(join!=nil?"/tmp/":nil):workfilename,true)))
                                continue;

                if (lockflag && badfile == 0) {
                        /* rewrite the rest of the RCSfile */
                        fastcopy(finptr,frewrite);
                        ffclose(frewrite); frewrite=NULL;
			ignoreints();
                        if (rename(newRCSfilename,RCSfilename)<0) {
                                error("Can't rewrite %s; saved in: %s",
                                RCSfilename, newRCSfilename);
                                newRCSfilename[0]='\0'; /* avoid deletion*/
				catchints();
                                break;
                        }
                        newRCSfilename[0]='\0'; /* avoid re-deletion by cleanup()*/
                        if (chmod(RCSfilename,RCSstat.st_mode & ~0222)<0)
                            warn("Can't preserve mode of %s",RCSfilename);
			catchints();
                }

#               ifdef SNOOPFILE
                logcommand("co",targetdelta,gendeltas,caller);
#               endif

                if (join) {
                        rmsema(); /* kill semaphore file so other co's can proceed */
                        if (!buildjoin(neworkfilename,tostdout)) continue;
                }
                if (!tostdout) {
                        if (link(neworkfilename,workfilename) <0) {
                                error("Can't create %s; see %s",workfilename,neworkfilename);
                                neworkfilename[0]= '\0'; /*avoid deletion*/
                                continue;
                        }
		}
        }
        if (!tostdout) {
	    struct stat sbuf;

            if (chmod(workfilename, WORKMODE(RCSstat.st_mode))<0)
                warn("Can't adjust mode of %s",workfilename);
	    if (usemodtime) {
		struct tm t;
		long mtime;
		struct stat workfilestat;
		char *cp = targetdelta->date;

		/* modify date so partime can parse it -- may need changing
		 * if DATEFORM changes */
		cp[8] = ' ';	
		cp[11] = ':';
		cp[14] = ':';
		if (partime(cp,&t) && (mtime = maketime(&t)) != 0L) {
		    if (stat(workfilename,&workfilestat) < 0)
			warn("Can't stat %s",workfilename);
		    fixtime(workfilename,workfilestat.st_atime,(time_t)mtime);
		}
		else
		    warn("Invalid date for revision %s of %s",
				targetdelta->num, workfilename);
		cp[8] = '.';
		cp[11] = '.';
		cp[14] = '.';
	    }
	}
        if (!tostdout) diagnose("done");
        } while (cleanup(),
                 ++argv, --argc >=1);

        exit(nerror!=0);

}       /* end of main (co) */


/*****************************************************************
 * The following routines are auxiliary routines
 *****************************************************************/

int rmoldfile(ofile)
char * ofile;
/* Function: unlinks ofile, if it exists, under the following conditions:
 * If the file is read-only, file is unlinked.
 * Otherwise (file writable):
 *   if !quietmode asks the user whether to really delete it (default: fail);
 *   otherwise failure.
 * Returns false on failure to unlink, true otherwise.
 */
{
        int response, c;    /* holds user response to queries */
        struct stat buf;

        if (stat (ofile, &buf) < 0)         /* File doesn't exist */
            return (true);                  /* No problem         */

        if (buf.st_mode & 0222) {            /* File is writable */
            if (!quietflag) {
                fprintf(stderr,"writable %s exists; overwrite? [ny](n): ",ofile);
                /* must be stderr in case of IO redirect */
                c=response=getchar();
                while (!(c==EOF || c=='\n')) c=getchar(); /*skip rest*/
                if (!(response=='y'||response=='Y')) {
                        warn("checkout aborted.");
                        return false;
                }
            } else {
                error("writable %s exists; checkout aborted.",ofile);
                return false;
            }
        }
        /* now unlink: either not writable, or permission given */
        if (unlink(ofile) != 0) {            /* Remove failed   */
            error("Can't unlink %s",ofile);
            return false;
        }
        return true;
}


creatempty(file)
char * file;
/* Function: creates an empty file named file.
 * Removes an existing file with the same name with rmoldfile().
 */
{
        int  fdesc;              /* file descriptor */

        if (!rmoldfile(file)) return false;
        fdesc=creat(file,0666);
        if (fdesc < 0) {
                faterror("Cannot create %s",file);
                return false;
        } else {
                close(fdesc); /* empty file */
                return true;
        }
}



/*****************************************************************
 * The rest of the routines are for handling joins
 *****************************************************************/

char * getrev(sp, tp, buffsize)
register char * sp, *tp; int buffsize;
/* Function: copies a symbolic revision number from sp to tp,
 * appends a '\0', and returns a pointer to the character following
 * the revision number; returns nil if the revision number is more than
 * buffsize characters long.
 * The revision number is terminated by space, tab, comma, colon,
 * semicolon, newline, or '\0'.
 * used for parsing the -j option.
 */
{
        register char c;
        register int length;

        length = 0;
        while (((c= *sp)!=' ')&&(c!='\t')&&(c!='\n')&&(c!=':')&&(c!=',')
                &&(c!=';')&&(c!='\0')) {
                if (length>=buffsize) return false;
                *tp++= *sp++;
                length++;
        }
        *tp= '\0';
        return sp;
}



int preparejoin()
/* Function: Parses a join list pointed to by join and places pointers to the
 * revision numbers into joinlist.
 */
{
        struct hshentry * (* joindeltas)[];
        struct hshentry * tmpdelta;
        register char * j;
        char symbolrev[revlength],numrev[revlength];

        joindeltas = (struct hshentry * (*)[])malloc(hshsize*sizeof(struct hshentry *));
        j=join;
        lastjoin= -1;
        for (;;) {
                while ((*j==' ')||(*j=='\t')||(*j==',')) j++;
                if (*j=='\0') break;
                if (lastjoin>=joinlength-2) {
                        error("too many joins");
                        return(false);
                }
                if(!(j=getrev(j,symbolrev,revlength))) return false;
                if (!expandsym(symbolrev,numrev)) return false;
                tmpdelta=genrevs(numrev,nil,nil,nil,joindeltas);
                if (tmpdelta==nil)
                        return false;
                else    joinlist[++lastjoin]=tmpdelta->num;
                while ((*j==' ') || (*j=='\t')) j++;
                if (*j == ':') {
                        j++;
                        while((*j==' ') || (*j=='\t')) j++;
                        if (*j!='\0') {
                                if(!(j=getrev(j,symbolrev,revlength))) return false;
                                if (!expandsym(symbolrev,numrev)) return false;
                                tmpdelta=genrevs(numrev,nil,nil,nil,joindeltas);
                                if (tmpdelta==nil)
                                        return false;
                                else    joinlist[++lastjoin]=tmpdelta->num;
                        } else {
                                error("join pair incomplete");
                                return false;
                        }
                } else {
                        if (lastjoin==0) { /* first pair */
                                /* common ancestor missing */
                                joinlist[1]=joinlist[0];
                                lastjoin=1;
                                /*derive common ancestor*/
                                joinlist[0]=malloc(revlength);
                                if (!getancestor(targetdelta->num,joinlist[1],joinlist[0]))
                                       return false;
                        } else {
                                error("join pair incomplete");
                                return false;
                        }
                }
        }
        if (lastjoin<1) {
                error("empty join");
                return false;
        } else  return true;
}



buildjoin(initialfile, tostdout)
char * initialfile; int tostdout;
/* Function: merge pairs of elements in joinlist into initialfile
 * If tostdout==true, copy result to stdout.
 * All unlinking of initialfile, rev2, and rev3 should be done by cleanup().
 */
{       char command[NCPPN+80];
        char subs[revlength];
        char * rev2, * rev3;
        int i;

        rev2=mktempfile("/tmp/",JOINFIL2);
        rev3=mktempfile("/tmp/",JOINFIL3);

        i=0;
        while (i<lastjoin) {
                /*prepare marker for merge*/
                if (i==0)
                        strcpy(subs,targetdelta->num);
                else    sprintf(subs, "merge%d",i/2);
                diagnose("revision %s",joinlist[i]);
                sprintf(command,"%s/co -p%s -q  %s > %s\n",TARGETDIR,joinlist[i],RCSfilename,rev2);
                if (system(command)) {
                        nerror++;return false;
                }
                diagnose("revision %s",joinlist[i+1]);
                sprintf(command,"%s/co -p%s -q  %s > %s\n",TARGETDIR,joinlist[i+1],RCSfilename,rev3);
                if (system(command)) {
                        nerror++; return false;
                }
                diagnose("merging...");
                sprintf(command,"%s %s%s %s %s %s %s\n", MERGE,
                        ((i+2)>=lastjoin && tostdout)?"-p ":"",
                        initialfile,rev2,rev3,subs,joinlist[i+1]);
                if (system(command)) {
                        nerror++; return false;
                }
                i=i+2;
        }
        return true;
}
